mlflow/models/model.py

```diff
@@ -747,9 +762,36 @@ def log(
                     await_registration_for=await_registration_for,
                     local_model_path=local_path,
                 )
-        model_info = mlflow_model.get_model_info()
-        if registered_model is not None:
-            model_info.registered_model_version = registered_model.version
+            model_info = mlflow_model.get_model_info()
+            if registered_model is not None:
+                model_info.registered_model_version = registered_model.version
+
+            # validate input example works for serving when logging the model
+            serving_input = mlflow_model.get_serving_input(local_path)
+            if mlflow_model.signature is None and serving_input is None:
+                _logger.warning(
+                    "Input example should be provided to infer model signature if the model "
+                    "signature is not provided when logging the model."
+                )
+            if serving_input:
+                from mlflow.pyfunc import validate_serving_input
+
+                try:
+                    validate_serving_input(model_info.model_uri, serving_input)
+                except Exception as e:
+                    _logger.warning(
+                        f"Failed to validate serving input example {serving_input}. "
```

Comment:
Having two examples (raw and serving) may be confusing. Do we need to explain what they are in the docs?

Reply:
We should do two things here: 
1. In the UI component (follow-up PR) display for code examples (pandas, spark_udf, and the new model serving-compatible input), utilize the legacy input_example for the pandas-based example and clearly articulate in that UI element why it may be different; similarly, for serving-compatible formatting, use the new structure added in this PR for the new example rendering and make it clear why this schema is structured as it is. 
2. Update https://mlflow.org/docs/latest/models.html#model-signatures-and-input-examples and https://mlflow.org/docs/latest/model/signatures.html in a 2nd follow up PR to explain (with screenshots) what each of these examples mean and why they are different. 

I think that should be sufficient to communicate to users to make this not confusing.

Reply:
While the current input_example.json is a bit confusing because it sometimes saves the raw input and sometimes converts it to a serving input, I'll clean it up to differentiate with the serving_input_example completely.

Reply:
Excellent :D